pkg/R/utils.R

```diff
@@ -3,56 +3,68 @@
-           function(index) {
-             obj <- callJMethod(jList, "get", as.integer(index))
+          function(index) {
+            obj <- callJMethod(jList, "get", as.integer(index))
 
-             # Assume it is either an R object or a Java obj ref.
-             if (inherits(obj, "jobj")) {
-               if (isInstanceOf(obj, "scala.Tuple2")) {
-                 # JavaPairRDD[Array[Byte], Array[Byte]].
+            # Assume it is either an R object or a Java obj ref.
+              if (inherits(obj, "jobj")) {
+                if (isInstanceOf(obj, "scala.Tuple2")) {
+                  # JavaPairRDD[Array[Byte], Array[Byte]].
 
-                 keyBytes = callJMethod(obj, "_1")
-                 valBytes = callJMethod(obj, "_2")
-                 res <- list(unserialize(keyBytes),
-                             unserialize(valBytes))
-               } else {
-                 stop(paste("utils.R: convertJListToRList only supports",
-                            "RDD[Array[Byte]] and",
-                            "JavaPairRDD[Array[Byte], Array[Byte]] for now"))
-               }
-             } else {
-               if (inherits(obj, "raw")) {
-                 # RDD[Array[Byte]]. `obj` is a whole partition.
-                 res <- unserialize(obj)
-                 # For serialized datasets, `obj` (and `rRaw`) here corresponds to
-                 # one whole partition dense-packed together. We deserialize the
-                 # whole partition first, then cap the number of elements to be returned.
-
-                 # TODO: is it possible to distinguish element boundary so that we can
-                 # unserialize only what we need?
-                 if (!is.null(logicalUpperBound)) {
-                   res <- head(res, n = logicalUpperBound)
+                  keyBytes = callJMethod(obj, "_1")
+                  valBytes = callJMethod(obj, "_2")
+                  res <- list(unserialize(keyBytes),
+                    unserialize(valBytes))
+                  } else {
+                    stop(paste("utils.R: convertJListToRList only supports",
+                      "RDD[Array[Byte]] and",
+                      "JavaPairRDD[Array[Byte], Array[Byte]] for now"))
                  }
-               } else {
-                 # obj is of a primitive Java type, is simplified to R's
-                 # corresponding type.
-                 res <- list(obj)
-               }
-             }
-             res
-           })
+            } else {
+              if (inherits(obj, "raw")) {
+                if (serialized == "byte") {
+                  # RDD[Array[Byte]]. `obj` is a whole partition.
+                  res <- unserialize(obj)
+                  # For serialized datasets, `obj` (and `rRaw`) here corresponds to
+                  # one whole partition dense-packed together. We deserialize the
+                  # whole partition first, then cap the number of elements to be returned.
+                  } else {
```

Comment:
Turns out the formatting on this function was just a mess after the merge. I've incorporated your change but also tried to straighten out the formatting across the whole thing, so hopefully it'll look better overall.